0.5/um_cmd/mstack.1
.\" Copyright (c) 2007 Renzo Davoli
.\"
.\" mstack manual page
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.

.TH VIEW-OS 1 "April 23, 2008" "VIEW-OS: a process with a view"
.SH NAME
mstack \- networking stack selection
.SH SYNOPSIS
.B mstack
[\fI-u46npbihv\fR]
[\fI-f\fR num,num#]
\fIstack_mountpoint\fR \fIcommand\fI
.br
.SH DESCRIPTION
.PP
In View-OS multiple networking stacks are available to processes
at the same time. View-OS supports the \fBmsocket(2)\fR system call.
mstack defines the default stack (for each address family) that is used
when msocket refers to a NULL pathname for its stack.
mstack provides also a backward compatibility for all the programs
using the obsolete \fBsocket(2)\fR system call.
The stack mountpoint identifies the stack. In ViewOS it is the mountpoint
where the stack was mounted (by \fBmount(2)(8)\fR.
In fact the call:
.br
"socket(domain,type,protocol)" 4
.br
is equivalent to:
.br
"msocket(NULL,domain,type protocol)" 4
.br
mstacks changes the default stack for the desired protocol families and
executes the command (by \fBexecve(2)).
.SH OPTIONS
.IP "\fB\-h\fR" 4
prints the mstack command usage
.IP "\fB\-v\fR" 4
sets the verbose mode on
.IP "\fB\-u\fR" 4
redefines the protocol stack for AF_UNIX \fBunix(7)\fR
.IP "\fB\-4\fR" 4
redefines the protocol stack for AF_INET, i.e. ipv4 \fBip(7)\fR
.IP "\fB\-6\fR" 4
redefines the protocol stack for AF_INET6, i.e. ipv6 \fBipv6(7)\fR
.IP "\fB\-n\fR" 4
redefines the protocol stack for AF_NETLINK \fBnetlink(7)\fR
.IP "\fB\-p\fR" 4
redefines the protocol stack for AF_PACKET \fBpacket(7)\fR
.IP "\fB\-b\fR" 4
redefines the protocol stack for AF_BLUETOOTH\fR
.IP "\fB\-i\fR" 4
redefines the protocol stack for AF_IRDA\fR
.IP "\fB\-f\fR \fInum,num...\fR" 4
redefines the protocol stack for the listed protocols (list of numeric tags).
.SH EXAMPLES
.IP "\fBmstack /dev/net/lwip ip addr\fR"
.br
lists the address managed by the stack mounted on \fB/dev/net/lwip\fR.
.IP "\fBexec mstack /dev/net/null bash\fR"
.br
disables the standard stack in a new bash (which overcomes the current one).
.SH SEE ALSO
.BR umview(1), 
.BR	kmview(1), 
.BR	msocket(3), 
.BR umnetnative(9),
.BR umnetnull(9),
.BR umnetlwipv6(9),
.BR	unix(7), 
.BR	ip(7), 
.BR	ipv6(7), 
.BR	netlink(7), 
.BR	packet(7).
.SH AUTHORS
View-OS is a project of the Computer Science Department, University of
Bologna. Project Leader: Renzo Davoli. 
.br
<http://www.sourceforge.net/projects/view-os>

Howto's and further information can be found on the project wiki
<wiki.virtualsquare.org>.



0.1/NEWS
Bug fixed in fat_write_data
Bug fixed in fat_mkdir (zeroing of new allocated clusters)

xmview-os/um_cmd/mstack.c
/*   This is part of um-ViewOS
 *   The user-mode implementation of OSVIEW -- A Process with a View
 *
 *   mstack: default stack choice
 *   
 *   Copyright 2005 Renzo Davoli University of Bologna - Italy
 *   
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License, version 2, as
 *   published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA.
 *
 *   $Id: um_add_service.c 362 2007-06-08 14:31:54Z rd235 $
 *
 */   
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <config.h>
#include <libgen.h>
#include <msocket.h>

#define AF_MAXMAX AF_MAX+2

int all=1;
int verbose=0;
char family[AF_MAXMAX+1];
char *fname[AF_MAXMAX+1]={
"PF_UNSPEC",
"PF_UNIX ",
"PF_INET",
"PF_AX25",
"PF_IPX",
"PF_APPLETALK",
"PF_NETROM",
"PF_BRIDGE",
"PF_ATMPVC",
"PF_X25",
"PF_INET",
"PF_ROSE",
"PF_DECnet",
"PF_NETBEUI",
"PF_SECURITY",
"PF_KEY",
"PF_NETLINK",
"PF_PACKET ",
"PF_ASH",
"PF_ECONET",
"PF_ATMSVC",
"PF_PROTO21",
"PF_SNA",
"PF_IRDA",
"PF_PPPOX",
"PF_WANPIPE",
"PF_PROTO26",
"PF_PROTO27",
"PF_PROTO28",
"PF_PROTO29",
"PF_PROTO30",
"PF_BLUETOOTH",
"PF_PROTO32",
"PF_PROTO33",
"PF_PROTO34"};

void usage()
{
	fprintf(stderr, "Usage:\n\tmstack [-u46npbihv] [-f #] stack_mountpoint command\n");
	exit(0);
}

main(int argc, char *argv[])
{
	int c;
	argv++;
	argc--;
	while (**argv == '-') {
		char *s=*argv;
		char ff=0;
		while (*s) {
			switch (*s) {
				case '-' : break;
				case 'u' : all=0;family[AF_UNIX]=1; break;
				case '4' : all=0;family[AF_INET]=1; break;
				case '6' : all=0;family[AF_INET6]=1; break;
				case 'n' : all=0;family[AF_NETLINK]=1; break;
				case 'p' : all=0;family[AF_PACKET]=1; break;
				case 'b' : all=0;family[AF_BLUETOOTH]=1; break;
				case 'i' : all=0;family[AF_IRDA]=1; break;
				case 'f' : all=0;ff=1;break;
				case 'v' : verbose=1;break;
				case 'h'	: usage();break;
				default: 
										fprintf(stderr,"Unknown option %c\n",*s);
										usage();
			}
			s++;
		}
		if (ff) {
			argv++;
			argc--;
			s=*argv;
			while (*s) {
				int fam=atoi(s);
				if (fam > 0 && fam <= AF_MAXMAX) 
					family[fam]=1;
				while (*s >= '0' && *s <= '9')
					s++;
				if (*s == ',')
					s++;
			}
		}
		argv++;
		argc--;
	}

	if (argc < 2)
		usage();
	else {
		int fd;
		if (all) {
			if (verbose)
				fprintf(stderr, "pid %d: stack %s for all supported protocol families\n",getpid(),argv[0]);
			if ((fd=msocket(argv[0],0,SOCK_DEFAULT,0)) < 0) {
				perror("mstack");
				exit(-1);
			}
		}	else {
			int i;
			for (i=1;i<=AF_MAXMAX;i++)
				if (family[i]) {
					if (verbose)
						fprintf(stderr, "pid %d: stack %s for %d=%s\n",getpid(),argv[0],i,fname[i]);
					if ((fd=msocket(argv[0],i,SOCK_DEFAULT,0)) < 0) {
						perror("mstack");
						exit(-1);
					}
				}
		}
		argv ++;
		argc --;
		execvp(basename(argv[0]),argv);
	}
}
